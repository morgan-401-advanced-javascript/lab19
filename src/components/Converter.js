import React from 'react';
import '../styles/styles.scss';
/**
 * Converter Class that will convert user input value from standard measurment units to metric units
 */
class Converter extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            x: this.props.x ? this.props.x : '',
            y: this.props.y ? this.props.y : '',
            rate: this.props.rate ? this.props.rate : 1,
        };
    }
/**
 * @method roundIt 
 * @param {number} num integer
 */
    roundIt(num) {
        return Math.round(num * 1000) / 1000;
    }
/**
 * universal change handler function
 * @method handleChange
 * @param {object} e 
 */
    handleChange(e) {
        if (!e.target.value) {
            this.setState({ x: '', y: '' });
            return;
        }

        let value = parseFloat(e.target.value);
        if (e.target.id === 'x')
            this.setState({
                x: value,
                y: this.roundIt(value * this.state.rate),
            });
        else if (e.target.id === 'y')
            this.setState({
                x: this.roundIt(value / this.state.rate),
                y: value,
            });
    }
/**
 * This will render the HTML onto the site
 * @method render
 */
    render() {
        let xLabel = this.props.xLabel ? this.props.xLabel : '1';
        let yLabel = this.props.yLabel || '1';
        return (
            <div className='converter'>
                <h4 className='converter-heading'>
                    Converting {xLabel} to/from {yLabel}
                </h4>

                <label className='x-label'>
                    <input
                        type='number'
                        step='any'
                        id='x'
                        value={this.state.x}
                        onChange={this.handleChange.bind(this)}
                    />
                    {xLabel}
                </label>
                <span className='mid-label'> is equal to </span>
                <label className='y-label'>
                    <input
                        type='number'
                        step='any'
                        id='y'
                        value={this.state.y}
                        onChange={this.handleChange.bind(this)}
                    />

                    {yLabel}
                </label>
            </div>
        );
    }
}

export default Converter;
