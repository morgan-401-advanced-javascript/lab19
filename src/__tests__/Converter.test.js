import React from 'react';
import renderer from 'react-test-renderer';
import Converter from '../components/Converter';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';


configure({ adapter: new Adapter() });

describe('', () => {
    it('', () => {
        expect(true).toBeTruthy();
    });
});

// Snapshot test
describe('Converter component', () => {
    it('renders correctly', () => {
        const page = renderer.create(<Converter />).toJSON();
        expect(page).toMatchSnapshot();
    });
});

describe('properly changes state', () => {
    it('Test that changing the x input field properly changes the state', () => {
        let component = mount(<Converter rate={1.5}/>);
        let xInput = component.find('input#x');
        xInput.simulate('change', {target: {id: 'x', value: 1}});
        expect(component.state('y')).toBe(1.5);
    });
    it('Test that changing the y input field properly changes the state', () => {
        let component = mount(<Converter rate={1.5}/>);
        let yInput = component.find('input#y');
        yInput.simulate('change', {target: {id: 'y', value: 3}});
        expect(component.state('x')).toEqual(2);
    });
    });


describe('Test that the component renders two input field', () => {
    it('Renders the first input fields', () => {
        let component = mount(<Converter />);
        expect(component.find('input#x').exists()).toEqual(true);
    });
    it('Renders the second input field', () => {
        let component = mount(<Converter />);
        expect(component.find('input#y').exists()).toEqual(true);
    });
    });
// Test that the labels are correctly set
describe('Test that the labels are correctly set', () => {
    it('The first label is correct ', () => {
        let component = mount(<Converter />);
        expect(component.find('label.x-label').text()).toBe(
            '1',
        );
    });
    it('The second label is correct ', () => {
        let component = mount(<Converter />);
        expect(component.find('label.y-label').text()).toBe(
            '1',
        );
    });
});
