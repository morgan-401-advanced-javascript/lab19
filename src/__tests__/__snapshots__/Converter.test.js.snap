// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Converter component renders correctly 1`] = `
<div
  className="converter"
>
  <h4
    className="converter-heading"
  >
    Converting 
    1
     to/from 
    1
  </h4>
  <label
    className="x-label"
  >
    <input
      id="x"
      onChange={[Function]}
      step="any"
      type="number"
      value=""
    />
    1
  </label>
  <span
    className="mid-label"
  >
     is equal to 
  </span>
  <label
    className="y-label"
  >
    <input
      id="y"
      onChange={[Function]}
      step="any"
      type="number"
      value=""
    />
    1
  </label>
</div>
`;
